cmd/skopeo/proxy.go

```diff
@@ -0,0 +1,359 @@
+	if err != nil {
+		return fmt.Errorf("failed to load image: %w", err)
+	}
+	h.img = &img
+	h.imgsrc = &imgsrc
+	return nil
+}
+
+func (h *proxyHandler) GetManifest(args []interface{}) (replyBuf, error) {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	var ret replyBuf
+
+	if len(args) != 0 {
+		return ret, fmt.Errorf("invalid request, expecting zero arguments")
+	}
+
+	if err := h.ensureImage(); err != nil {
+		return ret, err
+	}
+
+	ctx := context.TODO()
+	rawManifest, _, err := (*h.img).Manifest(ctx)
+	if err != nil {
+		return ret, err
+	}
+	digest, err := manifest.Digest(rawManifest)
+	if err != nil {
+		return ret, err
+	}
+	ociManifest, err := manifest.OCI1FromManifest(rawManifest)
+	if err != nil {
+		return ret, err
+	}
+	ociSerialized, err := ociManifest.Serialize()
+	if err != nil {
+		return ret, err
+	}
+
+	piper, pipew, err := os.Pipe()
+	if err != nil {
+		return ret, err
+	}
+	f := activePipe{
+		w: pipew,
+	}
+	h.activePipes[uint32(pipew.Fd())] = &f
+	f.wg.Add(1)
+	go func() {
+		// Signal completion when we return
+		defer f.wg.Done()
+		_, err = io.Copy(f.w, bytes.NewReader(ociSerialized))
+		if err != nil {
+			f.err = err
+		}
+	}()
+
+	r := replyBuf{
+		value:  digest.String(),
+		fd:     piper,
+		pipeid: uint32(pipew.Fd()),
+	}
+	return r, nil
+}
+
+func (h *proxyHandler) GetBlob(args []interface{}) (replyBuf, error) {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	var ret replyBuf
+
+	if len(args) != 1 {
+		return ret, fmt.Errorf("invalid request, expecting one blobid")
+	}
+
+	digestStr, ok := args[0].(string)
+	if !ok {
+		return ret, fmt.Errorf("expecting string blobid")
+	}
+
+	if err := h.ensureImage(); err != nil {
+		return ret, err
+	}
+
+	piper, pipew, err := os.Pipe()
+	if err != nil {
+		return ret, err
+	}
+
+	ctx := context.TODO()
+	d, err := digest.Parse(digestStr)
+	if err != nil {
+		return ret, err
+	}
+	blobr, blobSize, err := (*h.imgsrc).GetBlob(ctx, types.BlobInfo{Digest: d, Size: -1}, h.cache)
+	if err != nil {
+		return ret, err
+	}
+
+	f := activePipe{
+		w: pipew,
+	}
+	h.activePipes[uint32(pipew.Fd())] = &f
+
+	f.wg.Add(1)
+	go func() {
+		// Signal completion when we return
+		defer f.wg.Done()
+		verifier := d.Verifier()
+		tr := io.TeeReader(blobr, verifier)
+		_, err = io.Copy(f.w, tr)
+		if err != nil {
+			f.err = err
+			return
+		}
+		if !verifier.Verified() {
+			f.err = fmt.Errorf("corrupted blob, expecting %s", d.String())
+		}
+	}()
+
+	ret.value = blobSize
+	ret.fd = piper
+	ret.pipeid = uint32(pipew.Fd())
+	return ret, nil
+}
+
+func (h *proxyHandler) FinishPipe(args []interface{}) (replyBuf, error) {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	var ret replyBuf
+
+	pipeidf, ok := args[0].(float64)
+	if !ok {
+		return ret, fmt.Errorf("finishpipe: expecting blobid, not %T", args[0])
+	}
+	pipeid := uint32(pipeidf)
+
+	f, ok := h.activePipes[pipeid]
+	if !ok {
+		return ret, fmt.Errorf("finishpipe: no active pipe %d", pipeid)
+	}
+
+	f.wg.Wait()
+	f.w.Close()
```

Comment:
Having the server just close the pipe is the obvious thing to do indeed, and that's how it worked at first.  The reason I switched is because doing it this way forces the caller to also check for errors.

In the case of GetManifest, the only thing that can fail is the server writing to the client's pipe, which is kind of circular I guess.  But the goal is to ensure clients check the error from `GetBlob`.